### PR TITLE
Add meme_chip

### PR DESCRIPTION
**Pull request by gregvonkuster**

### CODE REVIEW

**Comment by gregvonkuster** [comment]
ping @bgruening

File: tools/meme_chip/meme_chip.xml [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+                        <option value="yes" selected="true">Yes</option>
+                        <option value="no">No</option>
+                    </param>
+                    <when value="yes">
+
+                        We have 2 dynamic select lists here.  The first select list (meme_motif_database_dir) is populated from the meme_motif_databases
+                        data table.  The second select list (meme_motif_database) is dynamically re-rendered whenever the selection in the meme_motif_database_dir
+                        select list is changed.  This composition used to work (see Examples->Dynamic Options section of
+                        https://docs.galaxyproject.org/en/latest/dev/schema.html) but no longer does.  We'll have to figure out what is broken in
+                        the dynamic options code in ~/parameters/basic.py in order to uncomment this block.
```
Comment (bgruening) [inline tools/meme_chip/meme_chip.xml]: @gregvonkuster have you added `<code file="get_meme_motif_databases.py" />` to the tool before testing this?

Comment (gregvonkuster) [reply tools/meme_chip/meme_chip.xml]: Yes, I had that tag set, but obviously it got lost in the shuffle before I submitted this PR.   Sorry about that! :(  I've just re-added it.

**Comment by gregvonkuster** [comment]
Here is the 2nd select tag for the tool.

```
<param name="meme_motif_database" type="select" label="MEME motif database" dynamic_options="get_meme_motif_database_options(file_path=meme_motif_database_dir)"/>
```

With the above setting (which is what should work), the tool for displays this...

options

If I hard-code the location of the ARABD database in the tag like this.

```<param name="meme_motif_database" type="select" label="MEME motif database" dynamic_options="get_meme_motif_database_options(file_path=/home/greg/work/git_workspace/galaxy/tool-data/meme_motif_databases/ARABD)"/>
```
...then the select list is populated correctly for that particular database.

options

I think this demonstrates there is a bug in the dynamic options for this particular category of associated dynamic select lists.

File: tools/meme_chip/get_meme_motif_databases.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import os
+
+
+def get_meme_motif_database_options(file_path):
+    options = []
+    if not os.path.isdir(file_path):
+        return options
+    for i, file_name in enumerate(os.listdir(file_path)):
+        full_path = os.path.join(file_path, file_name)
+        options.append((file_name, full_path, i == 0))
```
Comment (bgruening) [inline tools/meme_chip/get_meme_motif_databases.py]: What does this i == 0 do? Are you relying here on the oder of the returned os.listdir() ?

File: tools/meme_chip/meme_chip.xml [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements" />
+    <code file="get_meme_motif_databases.py" />
+    <command detect_errors="exit_code"><![CDATA[
+#import os
+#set primary_output = $os.path.join($output.files_path, "index.html")
+#set options_type = $options_type_cond.options_type
+meme-chip $input
```
Comment (bgruening) [inline tools/meme_chip/meme_chip.xml]: quotes

File: tools/meme_chip/meme_chip.xml [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+-noecho
+#if str($control) not in ['', 'None']:
```
Comment (bgruening) [inline tools/meme_chip/meme_chip.xml]: mh, `if control:` should be enough.

File: tools/meme_chip/meme_chip.xml [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    -neg '$control'
+#end if
+#if $sequence_alphabet == 'dna':
```
Comment (bgruening) [inline tools/meme_chip/meme_chip.xml]: You can replace the entire if/else with `$sequence_alphabet` isn't it? See below ...

File: tools/meme_chip/meme_chip.xml [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    exit 1;
+fi
+&& rm '$output'
+&& ln -s $primary_output '$output'
+    ]]></command>
+    <inputs>
+        <param name="input" type="data" format="fasta" label="Primary sequences" help="Nucleotide sequences must have equal length"/>
+        <param name="control" type="data" format="fasta" optional="true" label="Control sequences" help="If no selection, positive sequences in the input are shuffled to create the negative set"/>
+        <param name="sequence_alphabet" type="select" label="Sequence alphabet">
+            <option value="dna" selected="true">DNA</option>
```
Comment (bgruening) [inline tools/meme_chip/meme_chip.xml]: value="-dna" and value="-rna"

File: tools/meme_chip/meme_chip.xml [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    ##         -centrimo-ethresh $options_type_cond.run_centrimo_cond.centrimo_ethresh
+    ##     #end if
+    ##     #if str($options_type_cond.run_centrimo_cond.centrimo_noseq) == "true":
+    ##         -centrimo-noseq
+    ##     #end if
+    ##     #if str($options_type_cond.run_centrimo_cond.centrimo_flip) == "true":
+    ##         -centrimo-flip
+    ##     #end if
+    ## #end if
+    #if str($options_type_cond.search_given_strand) == "true":
```
Comment (bgruening) [inline tools/meme_chip/meme_chip.xml]: for booleans it should be enough to just ask `if $options_type_cond.search_given_strand:`

File: tools/meme_chip/meme_chip.xml [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    #if str($options_type_cond.meme_maxw) != "0":
+        -meme-maxw $options_type_cond.meme_maxw
+    #end if
+    #if str($options_type_cond.meme_nmotifs) != '0':
+        -meme-nmotifs $options_type_cond.meme_nmotifs
+    #end if
+    #if str($options_type_cond.meme_minsites) != "0":
+        -meme-minsites $options_type_cond.meme_minsites
+    #end if
+    #if str($options_type_cond.meme_maxsites) != "0":
```
Comment (bgruening) [inline tools/meme_chip/meme_chip.xml]: you could make all these params `optional=true` and just ask here `#if $options_type_cond.meme_maxsites`. the value from the params can then be removed as well I guess...

Comment (bgruening) [reply tools/meme_chip/meme_chip.xml]: Same for the above ...

File: tools/meme_chip/meme_chip.xml [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    #end if
+    #if str($options_type_cond.meme_pal) == "true":
+        -meme-pal
+    #end if
+    -dreme-e $options_type_cond.dreme_e
+    -dreme-m $options_type_cond.dreme_m
+    -spamo-skip
+    -fimo-skip
+#end if
+&>meme_chip_log.txt;
```
Comment (bgruening) [inline tools/meme_chip/meme_chip.xml]: could you not in any case output this to sdtout and remove the if condition?

Comment (gregvonkuster) [reply tools/meme_chip/meme_chip.xml]: @bgruening The intent for this is to check the status code and display the error if it is not 0.  I'll remove this per your request here, but I don't see why my approach is a problem.  Can you clarify?

File: tools/meme_chip/meme_chip.xml [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+                    <when value="yes">
+                        <param name="seed" type="integer" optional="true" value="0" min="0" label="Seed for the randomized selection of sequences" help="Zero value indicates random seeding"/>
+                    </when>
+                    <when value="no"/>
+                </conditional>
+                <param name="ccut" type="integer" optional="true" value="100" min="0" label="maximum size of a sequence before it is cut down to a centered section" help="Zero value indicates the sequences should not be cut down"/>
+                <param name="group_threash" type="float" value="0.05" min="0" label="Primary threshold for clustering motifs" />
+                <param name="group_weak" type="float" optional="true" value="0" min="0" label="Secondary threshold for clustering motifs" help="Zero value results in 2*primary threshold"/>
+                <param name="filter_thresh" type="float" value="0.05" min="0" label="E-value threshold for including motifs"/>
+                <param name="search_given_strand" type="boolean" truevalue="true" falsevalue="" checked="False" label="Search given strand only"/>
```
Comment (nsoranzo) [inline tools/meme_chip/meme_chip.xml]: You can set `truevalue="-norc"` and simplify the handling of this param in the `<command>`.

File: tools/meme_chip/meme_chip.xml [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+#set primary_output = $os.path.join($output.files_path, "index.html")
+#set options_type = $options_type_cond.options_type
+meme-chip '$input'
+-noecho
+#if $control:
+    -neg '$control'
+#end if
+$sequence_alphabet
+-o '$output.files_path'
+#if str($options_type)=='advanced':
```
Comment (nsoranzo) [inline tools/meme_chip/meme_chip.xml]: This should be `$options_type_cond.options_type`

File: tools/meme_chip/meme_chip.xml [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+                </conditional>
+                -->
+                <param name="background_model_order" type="select" label="Select the order of the Markov background model">
+                    <option value="0">0-order model of sequences</option>
+                    <option value="1" selected="True">1st order model of sequences</option>
+                    <option value="2">2nd order model of sequences</option>
+                    <option value="3">3rd order model of sequences</option>
+                    <option value="4">4th order model of sequences</option>
+                </param>
+                <param name="nmeme" type="integer" optional="true" value="0" min="0" label="Limit of sequences to pass to MEME" help="Zero value has no effect"/>
```
Comment (nsoranzo) [inline tools/meme_chip/meme_chip.xml]: I'd use `min="1" value=""` and remove the `help`.

File: tools/meme_chip/meme_chip.xml [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+                <conditional name="subsampling_cond">
+                    <param name="subsampling" type="select" label="Should subsampling be random?" help="Select 'No' if your input sequences are sorted in order of confidence (best to worst)">
+                        <option value="yes" selected="true">Yes</option>
+                        <option value="no">No</option>
+                    </param>
+                    <when value="yes">
+                        <param name="seed" type="integer" optional="true" value="0" min="0" label="Seed for the randomized selection of sequences" help="Zero value indicates random seeding"/>
```
Comment (nsoranzo) [inline tools/meme_chip/meme_chip.xml]: I'd use `min="1" value=""` and remove the `help`.

File: tools/meme_chip/meme_chip.xml [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+    ##     #end if
+    ## #end if
+    #if $options_type_cond.search_given_strand:
+        -norc
+    #end if
+    -order $options_type_cond.background_model_order
+    #if str($options_type_cond.subsampling_cond.subsampling) == "no":
+        -norand
+        #set seed = $options_type_cond.subsampling_cond.subsampling.seed
+        #if $seed:
```
Comment (nsoranzo) [inline tools/meme_chip/meme_chip.xml]: You can merge the 2 lines above in:
```
        #if $options_type_cond.subsampling_cond.subsampling.seed
```

File: tools/meme_chip/meme_chip.xml [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        #end if
+    #end if
+    #if $options_type_cond.nmeme:
+        -nmeme $options_type_cond.nmeme
+    #end if
+    #if $options_type_cond.ccut:
+        -ccut $options_type_cond.ccut
+    #end if
+    -group-thresh $options_type_cond.group_threash
+    #if $options_type_cond.group_weak:
```
Comment (nsoranzo) [inline tools/meme_chip/meme_chip.xml]: If 0 is an allowed value for `$options_type_cond.group_weak`, this should be `#if str($options_type_cond.group_weak):`

File: tools/meme_chip/meme_chip.xml [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+                    </when>
+                    <when value="no"/>
+                </conditional>
+                <param name="ccut" type="integer" optional="true" value="100" min="0" label="maximum size of a sequence before it is cut down to a centered section" help="Zero value indicates the sequences should not be cut down"/>
+                <param name="group_threash" type="float" value="0.05" min="0" label="Primary threshold for clustering motifs" />
+                <param name="group_weak" type="float" optional="true" value="0" min="0" label="Secondary threshold for clustering motifs" help="Zero value results in 2*primary threshold"/>
+                <param name="filter_thresh" type="float" value="0.05" min="0" label="E-value threshold for including motifs"/>
+                <param name="search_given_strand" type="boolean" truevalue="true" falsevalue="" checked="False" label="Search given strand only"/>
+                <param name="old_clustering" type="boolean" truevalue="true" falsevalue="" checked="False" label="Pick cluster seed motifs based only on significance"/>
```
Comment (nsoranzo) [inline tools/meme_chip/meme_chip.xml]: Replace `name="old_clustering"` with `argument="-old_clustering"` .
Also, you can set truevalue="-old_clustering" and simplify the handling of this param in the <command>.

File: tools/meme_chip/meme_chip.xml [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+                    <option value="oops" selected="True">One occurance per sequence</option>
+                    <option value="zoops">Zero or one occurances per sequence</option>
+                    <option value="anr">Any number of repititions</option>
+                </param>
+                <param name="meme_minw" type="integer" optional="true" value="0" min="0" label="Minimum motif width"/>
+                <param name="meme_maxw" type="integer" optional="true" value="0" min="0" label="Maximum motif width"/>
+                <param name="meme_nmotifs" type="integer" optional="true" value="0" min="0" label="Maximum number of motifs to find"/>
+                <param name="meme_minsites" type="integer" optional="true" value="0" min="0" label="Minimum number of sites per motif"/>
+                <param name="meme_maxsites" type="integer" optional="true" value="0" label="Maximum number of sites per motif"/>
+                <param name="meme_pal" type="boolean" truevalue="true" falsevalue="" checked="False" label="Look for palindromes only"/>
```
Comment (nsoranzo) [inline tools/meme_chip/meme_chip.xml]: As for `old_clustering`

**Comment by nsoranzo** [comment]
Thanks @gregvonkuster !

**Pull request merged**